server/src/main/java/org/elasticsearch/common/util/concurrent/EsExecutors.java

```diff
@@ -389,32 +441,23 @@ public boolean isSystem() {
      */
     private EsExecutors() {}
 
-    static class ExecutorScalingQueue<E> extends LinkedTransferQueue<E> {
+    private static <E> LinkedTransferQueue<E> newUnboundedScalingLTQueue(int corePoolSize, int maxPoolSize) {
+        // scaling beyond core pool size using an unbounded queue requires ExecutorScalingQueue
+        // note, reconfiguration of core / max pool size not supported in EsThreadPoolExecutor
+        return maxPoolSize > corePoolSize ? new ExecutorScalingQueue<>() : new LinkedTransferQueue<>();
+    }
 
-        ThreadPoolExecutor executor;
+    static class ExecutorScalingQueue<E> extends LinkedTransferQueue<E> {
```

Comment:
Could you add a detailed comment here describing the problem with just using `LinkedTransferQueue` everywhere? It's always felt pretty weird to just rewire all these methods into other methods on `super`, and this change makes it even more mysterious IMO. The method names are deeply unintuitive which doesn't help.

Likewise, why don't we need the workarounds for JDK-8323659 in this case?

Reply:
Granted, I'll add some Java docs explaining more in detail how the scaling behavior for ES thread pool is different from the standard JDK one 👍 Understanding the reasoning behind the old code took the largest part of the time here to be honest...

The workaround for JDK-8323659 is only needed if overriding `offer`.

Reply:
I've added fairly extensive Java docs, please let me know if that improves things @DaveCTurner

Reply:
++ thanks that helps